src/ripple/app/tx/impl/AMMWithdraw.cpp

```diff
@@ -372,19 +373,31 @@ AMMWithdraw::applyGuts(Sandbox& sb)
         return std::make_pair(tecINTERNAL, STAmount{});
     }();
 
-    // AMM is deleted if zero tokens balance
-    if (result == tesSUCCESS && newLPTokenBalance != beast::zero)
+    if (result != tesSUCCESS)
+        return {result, false};
+
+    bool updateBalance = true;
+    if (newLPTokenBalance == beast::zero)
+    {
+        if (auto const ter =
+                deleteAMMAccount(sb, ctx_.tx[sfAsset], ctx_.tx[sfAsset2], j_);
+            ter != tesSUCCESS && ter != tecINCOMPLETE)
+            return {ter, false};
+        else
+            updateBalance = (ter == tecINCOMPLETE);
+    }
+
+    if (updateBalance)
     {
         ammSle->setFieldAmount(sfLPTokenBalance, newLPTokenBalance);
         sb.update(ammSle);
-
-        JLOG(ctx_.journal.trace())
-            << "AMM Withdraw: tokens " << to_string(newLPTokenBalance.iou())
-            << " " << to_string(lpTokens.iou()) << " "
-            << to_string(lptAMMBalance.iou());
     }
 
-    return {result, result == tesSUCCESS};
+    JLOG(ctx_.journal.trace())
+        << "AMM Withdraw: tokens " << to_string(newLPTokenBalance.iou()) << " "
+        << to_string(lpTokens.iou()) << " " << to_string(lptAMMBalance.iou());
+
+    return {tesSUCCESS, true};
```

Comment:
if AMMWithdraw has delete functionality then it should behave the same way as AMMDelete, by returning tecIMCOMPLETE when there is more to delete.

If you are determined to code it in this way then I suggest only running deleteAMMAccount here if you know it will result in complete deletion otherwise don't run it at all here and let someone clean it up with AMMDelete.

Reply:
Returning `tecINCOMPLETE` makes sense.

We already went through the trouble of iterating over 1,500 trustlines, deleting them, and adjusting the owner count, so why through this effort away? It also results in fewer calls to AMMDelete.

Reply:
Mainly because the person who did the withdrawal doesn't care about deleting the AMM, they care about withdrawing their funds from the AMM. For every prior withdrawal until this final one, the withdrawer gets tesSUCCESS (i.e. withdrawal was successful.) The last one get tecINCOMPLETE. That's confusing as hell to them. Was the withdrawal incomplete? Overloading the functionality at this edge case requires losing information. Either the result code is consistent with withdraw or it's consistent with delete but it can't be both, unless it's a 100% delete every time.

Reply:
I think you are suggesting this. Please correct me if I'm wrong.

>        if last withdraw
>            if trustlines less than 1,500
>                delete trustlines
>                delete account
>                return tesSUCCESS
>            set LPTokens balance to 0
>            return tecINCOMPLETE

Is there an efficient way to get the number of account's trustlines?

Reply:
You could overload ownercount to count the number of incoming trustlines (since ownercount isn't really used in AMM?)
I would always return tesSUCCESS from AMMWithdrawal if the *withdrawal* was successful. Whatever ancillary clean-up activity the transactor also does is irrelevant to its return code and purpose. From the perspective of users and end developers they couldn't care less if the GC finished or not, all they want to know about is if the withdrawal was successful or not.

Reply:
I propose that in an AMM withdrawal, the withdrawal is the point. If that succeeds, the the transaction succeeds. Deleting trust lines is a helpful side effect. 

I suggest something like

* If withdraw fails
  * return appropriate tec code
* if last withdraw
  * delete up to X trustlines
  * if deletion fails
    * return appropriate tec code
  * if all trust lines deleted
    *  delete account
  * else
    * set LPTokens balance to 0
*   return tesSUCCESS

No need to change the meaning of OwnerCount

Reply:
This is not a good user experience from the withdrawer perspective. The withdrawer just want to withdraw the funds and doesn't care if the delete fails or not.

Reply:
I actually think `OwnerCount` is pretty useful in providing an estimate:
- Clean up scope for `AMMDelete` caller
- Number of Liquidity Providers

Reply:
I'm with Ed on this one, actually. I really don't like the idea that AMMWithdraw returns `tecINCOMPLETE` if it successfully withdrew funds.

IMO if an AMMWithdraw transaction withdraws the requested amount, it should return `tesSUCCESS`, full stop. Whether or not it also succeeded in fully cleaning up the AMM object is a detail that you can figure out from the metadata, but is not usually relevant to the user doing the withdrawal.

The meaning of AMMDelete is different so it makes sense that the semantics for `tecINCOMPLETE` should be different there. The "delete" transaction should return success if and only if the AMM was (fully) deleted. The `tecINCOMPLETE` result code is useful for telling you that it made progress on the thing you asked it to do, but isn't done; that's completely different than the Withdraw case where getting a `tec`-class code should always indicate that the withdraw operation didn't happen.